### PR TITLE
[FSDP] Fix a small bug of pre_backward_hook params prefetch

### PR DESCRIPTION
**Pull request by LinjianMa**
Fix a potential small bug in FSDP pre_backward_hook params prefetch.

In `_pre_backward_hook`, `self._need_prefetch_full_params(self.training_state)` is used to decide whether the params of the next backward pass needs to be pre-fetched, and currently it is also used to check whether we want to perform synchronization in the current backward pass before `_rebuild_full_params`.

For some edge cases, using this to check whether to perform synchronization is not current. One example is when `self._my_fsdp_idx_in_graph = 0`, which means this is the last backward pass. In this way, we have `self._need_prefetch_full_params(self.training_state)=False` since there is no backward pass after it, and currently synchronization will not be done before `_rebuild_full_params`.
But the params of this layer is prefetched at the previous layer, thus a synchronization needs to be done.

To fix this, we just needs to check whether to do the synchronization using another flag rather than `self._need_prefetch_full_params(self.training_state)`, and that is what this PR does.

### CODE REVIEW

File: torch/distributed/fsdp/fully_sharded_data_parallel.py [hunk 1]
```diff
@@ -2626,26 +2627,23 @@ def _pre_backward_hook(*unused: Any) -> None:
             if self._is_root:
                 self._queue_wait_for_post_backward()
 
-            if self._need_prefetch_pre_backward_hook():
```
Comment (LinjianMa) [inline torch/distributed/fsdp/fully_sharded_data_parallel.py:2629]: Note that `self._need_prefetch_pre_backward_hook()` decides whether the params of the **next backward layer** needs to be pre-fetched, so it does not make sense to use it to decide whether to perform synchronization or not.

One example is when `self._my_fsdp_idx_in_graph = 0`, under which case the if statement is false. But the params of this layer is prefetched, thus a synchronization needs to be done.

**Comment by LinjianMa** [comment]
Seems the changes breaks some tests, looking into that.

File: torch/distributed/fsdp/fully_sharded_data_parallel.py [hunk 2]
```diff
@@ -1455,6 +1455,7 @@ def _reset_lazy_init(self) -> None:
         self._streams: Dict[str, torch.cuda.Stream] = {}
         self._fsdp_graph_order: List[nn.Module] = []
         self._my_fsdp_idx_in_graph: Optional[int] = None
+        self._pre_backward_hook_full_params_prefetched: bool = False
```
Comment (zhaojuanmao) [inline torch/distributed/fsdp/fully_sharded_data_parallel.py:1458]: as commented in previous PR, flag is owned by its own layer only, so previous layer's flag can not be checked here

**Comment by zhaojuanmao** [comment]
by the way, let's add PR description for each PR as well

**Review by rohan-varma — COMMENTED** [review]
@LinjianMa @zhaojuanmao are we still planning to merge this one in?

**Comment by LinjianMa** [comment]
> @LinjianMa @zhaojuanmao are we still planning to merge this one in?

Hi @rohan-varma , sorry for not getting back to this for a while, since this one has a bit overlapping with the forward-prefetching, I will work on this soon once that one is merged

File: torch/distributed/fsdp/fully_sharded_data_parallel.py (hunk 2 above)
Comment (LinjianMa) [reply torch/distributed/fsdp/fully_sharded_data_parallel.py:1465]: Note that in line 2682, previous layer's _pre_backward_hook_full_params_prefetched flag is set, so in the forward of the previous layer this flag will be checked.

**Review by zhaojuanmao — APPROVED** [review]
nice catch! thanks for fixing this! let's run some benchmarks before landing, especially benchmarking for when backwardPrefetch is enabled

**Pull request closed without merging**